lib/wibox/container/conditional.lua

```diff
@@ -0,0 +1,362 @@
+--
+-- @property connected_global_signals
+-- @param table
+
+function conditional:get_connected_global_signals()
+    return self._private.connected_global_signals
+end
+
+function conditional:set_connected_global_signals(signals)
+    if self._private.connected_global_signals then
+        for _, v in pairs(self._private.connected_global_signals) do
+            v[1].disconnect_signal(v[2], self._private.eval_handler)
+        end
+    end
+
+    self._private.connected_global_signals = signals
+
+    for _, v in pairs(self._private.connected_global_signals) do
+        assert(v[1].connect_signal, "The object passed to `global_signals` is invalid")
+        assert(not v[1].is_widget,
+            "Regular objects cannot be used in `global_signals`, use `attached_signals`"
+        )
+        v[1].connect_signal(v[2], self._private.eval_handler)
+    end
+end
+
+--- Controls the arguments passed to the `when` callback.
+--
+-- This is useful when extending this container to fit specific purposes.
+--
+-- All values returned by this function will be passed to the `when` callback.
+-- Note that it will be called only once for all templates.
+--
+-- The default value returns the `conditional` widget instance.
+--
+-- The callback has a single parameter (the widget instance).
+--
+-- @property argument_callback
+-- @param function
+
+function conditional:set_argument_callback(callback)
+    self._private.argument_callback = callback
+end
+
+function conditional:get_argument_callback()
+    return self._private.argument_callback
+end
+
+--- Reset this layout. The widget will be removed and the rotation reset.
+function conditional:reset()
+    self:set_widget(nil)
+end
+
+--- Evaluate all the `when` clauses and update the current widget(s).
+--
+-- @DOC_wibox_container_conditional_eval_EXAMPLE@
+--
+-- The signal is `property::selected_template` and has the following argument
+--    * The `wibox.container.conditional` widget instance
+--    * The current template
+--    * The old template
+--    * The current template widget instance
+--    * The old template widget instance
+--
+function conditional:eval()
+
+    -- Avoid called eval multiple timer per event loop cycle due to the
+    -- attached signals. This creates and destroy widgets for nothing
+    self._private.exec_once = false
+
+    local old_selected, selected = self._private.selected ,nil
+
+    for _, v in ipairs(self._private.templates or {}) do
+        if v.when then
+            if v.when(self._private.argument_callback(self)) then
+                selected = v
+                break
+            end
+        elseif self._private.default_when then
+            local ret = self._private.default_when(v, self._private.argument_callback(self))
+
+            -- The function must return something or is considered broken
+            assert(type(ret) == "boolean", "The conditional layout `default_when`"..
+                " must return a boolean"
+            )
```

Comment:
Why? This feels a bit un-lua-y to me. Lua allows `if 42 then` while e.g. Java does not.